clients/src/main/java/org/apache/kafka/common/metrics/Sensor.java

```diff
@@ -97,7 +97,25 @@ public static RecordingLevel forName(String name) {
         public boolean shouldRecord(final int configId) {
             return configId == DEBUG.id || configId == this.id;
         }
+    }
 
+    public enum QuotaEnforcementType {
```

Comment:
I think that we do. I actually did exactly what you suggest in the first place but then realised that it introduces a race condition. For instance, two requests processed in two separate threads can verify the quota at the same time. If it is below the authorized quota, they would go ahead, record the value, and do the creation/deletion. That violates the strict enforcement of the quota as we don't want to accept a creation/deletion if the quota is already violated. In this case, the first request must succeed while the second must fail.

The proposed approach allows to do the quota verification and the recording within the sensor lock. That basically allows us to record the value iif the quota is not already violated.

This is probably less an issue for our existing quotas because we always accept requests, even if the quota is already violated.

Regarding the naming, we may find better ones. Would you have any suggestions?

Reply:
Thanks for the explanation, makes sense. 

Not sure about a better name, but it seems to me that we are changing a boolean `checkQuotas=true|false` to an enum that says `checkQuotas=false|before|after`. Perhaps something that makes that clearer.

Reply:
On one side, we have names that describe how the quota enforcement behave, and on the other side, we have names that describe where the quota is enforced. I personally lean towards the current proposal but I an open to discuss this. My reasoning is that the current proposal is aligned with the naming in the quota manager as well. That makes the overall more consistent IMO.